Source/JavaScriptCore/runtime/ObjectConstructor.cpp

```diff
@@ -324,10 +324,49 @@ JSC_DEFINE_HOST_FUNCTION(objectConstructorAssign, (JSGlobalObject* globalObject,
     // FIXME: Extend this for non JSFinalObject. For example, we would like to use this fast path for function objects too.
     // https://bugs.webkit.org/show_bug.cgi?id=185358
     bool targetCanPerformFastPut = jsDynamicCast<JSFinalObject*>(target) && target->canPerformFastPutInlineExcludingProto() && target->isStructureExtensible();
+    unsigned argsCount = callFrame->argumentCount();
+
+    if (argsCount > 2 && argsCount < 5 && targetCanPerformFastPut) {
+        bool willBatch = true;
+        for (unsigned i = 1; i < argsCount; ++i) {
+            JSValue sourceValue = callFrame->uncheckedArgument(i);
+            if (!sourceValue.isObject()) {
+                willBatch = false;
+                break;
+            }
+            JSObject* source = asObject(sourceValue);
+            if (!source->staticPropertiesReified() || !source->structure()->canPerformFastPropertyEnumeration()) {
+                willBatch = false;
+                break;
+            }
+        }
+        if (willBatch) {
+            Vector<RefPtr<UniquedStringImpl>, 16> properties;
+            MarkedArgumentBufferWithSize<16> values;
+            for (unsigned i = 1; i < argsCount; ++i) {
+                JSValue sourceValue = callFrame->uncheckedArgument(i);
+                JSObject* source = asObject(sourceValue);
+                source->structure()->forEachProperty(vm, [&](const PropertyTableEntry& entry) -> bool {
+                    if (entry.attributes() & PropertyAttribute::DontEnum)
+                        return true;
```

Comment:
nit: We should really change this to return `IterationStatus`, but we can do that later.  Let's see, true means `IterationStatus::Continue` here.